### PR TITLE
docker: multiple optimizations

### PR DESCRIPTION
**Pull request by crepererum**
See commits for details.

### CODE REVIEW

File: docs/developers/docker.rst [hunk 1]
```diff
@@ -142,6 +142,15 @@ As of this writing changing template files do not lead to application reloading
 and do not purge caches. As a workaround you can simple alter one of the python
 files, e.g. by using `touch`.
 
+.. note::
+    Changing the Python source files will invalidate stored bytecode files. For
+    security reasons, these bytecode files can only be recreatd by the root
```
Comment (jmartinm) [inline docs/developers/docker.rst]: typo: recreatd > recreated

File: invenio/base/factory.py [hunk 2]
```diff
@@ -258,7 +258,6 @@ def create_app(instance_path=None, static_folder=None, **kwargs_config):
             except (SyntaxError, ValueError):
                 pass
             app.config[cfg_name] = cfg_value
-            app.logger.debug("{0} = {1}".format(cfg_name, cfg_value))
```
Comment (jirikuncar) [inline invenio/base/factory.py]: This output is only present when `DEBUG=True`. Does it really bother so much?

Comment (crepererum) [inline invenio/base/factory.py]: For @tiborsimko and me it does.

**Comment by tiborsimko** [comment]
:+1: for suggested improvements and optimisations.

One comment only WRT independent Invenio packages such as `invenio-search` that are installed under `/code/src`:

1. Good that they are now installed as editable, this relieves read-only pain for developers.  However, the `/code/src` files are editable only to the `root` user, not to the regular `invenio` user.  Wanted?  I guess we could we make all packages under `/code/*` to be owned by the `invenio` user, so that people could more easily upgrade/downgrade/mount them with the default identity.

2. Cosmetics: what about putting `invenio` package at the same `/code/src` level, for easier multi-volume mounting of locally checked repositories in docker-compose?  (Kind of like we do for the overlay repository.)

3. Several usually non-editable dependencies such as `jinja2` are installed under `/code/src` too, rather than in the usual system wide `/usr/local/lib/python2.7/site-packages/` place.  Wanted?

Anyway, this could be part of later improvements, the current PR is already making life much easier for Docker-based installations.

File: requirements-devel.txt [hunk 3]
```diff
@@ -18,10 +18,13 @@
 -e git+git://github.com/inveniosoftware/invenio-client.git#egg=invenio-client
 -e git+git://github.com/inveniosoftware/invenio-collections.git#egg=invenio-collections
 -e git+git://github.com/inveniosoftware/invenio-comments.git#egg=invenio-comments
+-e git+git://github.com/inveniosoftware/invenio-deposit.git#egg=invenio-deposit
 -e git+git://github.com/inveniosoftware/invenio-documents.git#egg=invenio-documents
+-e git+git://github.com/inveniosoftware/invenio-groups.git#egg=invenio-groups
 -e git+git://github.com/inveniosoftware/invenio-oauthclient.git#egg=invenio-oauthclient
 -e git+git://github.com/inveniosoftware/invenio-pidstore.git#egg=invenio-pidstore
 -e git+git://github.com/inveniosoftware/invenio-records.git#egg=invenio-records
+-e git+git://github.com/inveniosoftware/invenio-schema.git#egg=invenio-jsonschemas
```
Comment (tiborsimko) [inline requirements-devel.txt]: Please rather use canonical `git+git://github.com/inveniosoftware/invenio-jsonschemas.git#egg=invenio-jsonschemas`.

File: Dockerfile [hunk 4]
```diff
@@ -29,24 +38,26 @@ RUN apt-get update && \
...
+        iojs \
         libffi-dev \
         liblzma-dev \
         libmysqlclient-dev \
         libssl-dev \
         libxslt-dev \
         mysql-client \
-        nodejs \
         poppler-utils \
-        subversion \
```
Comment (tiborsimko) [inline Dockerfile]: Note that if you remove subversion, the following tests would as-if-fail by wrong skipping: (note leading ERROR)

```
======================================================================
ERROR: test_harvest_with_tag_1 (invenio.testsuite.test_utils_vcs_svn.SVNHarvestTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/code/invenio/testsuite/test_utils_vcs_svn.py", line 43, in setUp
    raise SkipTest("SVN not found. It probably needs installing.")
SkipTest: SVN not found. It probably needs installing.

======================================================================
ERROR: test_harvest_with_tag_2 (invenio.testsuite.test_utils_vcs_svn.SVNHarvestTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/code/invenio/testsuite/test_utils_vcs_svn.py", line 43, in setUp
    raise SkipTest("SVN not found. It probably needs installing.")
SkipTest: SVN not found. It probably needs installing.

======================================================================
ERROR: test_harvest_without_tag (invenio.testsuite.test_utils_vcs_svn.SVNHarvestTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/code/invenio/testsuite/test_utils_vcs_svn.py", line 43, in setUp
    raise SkipTest("SVN not found. It probably needs installing.")
SkipTest: SVN not found. It probably needs installing.
```

Could properly decorate them, but this is a non-Docker issue, so can be attacked later.

File: Dockerfile [hunk 5]
```diff
@@ -29,24 +38,26 @@ RUN apt-get update && \
     apt-get -qy install --fix-missing --no-install-recommends \
         curl \
         && \
-    curl -sL https://deb.nodesource.com/setup_0.10 | bash -
+    curl -sL https://deb.nodesource.com/setup_iojs_2.x | bash -
```
Comment (tiborsimko) [inline Dockerfile]: I guess we should also amend `INSTALL.rst` and `docs/getting-started/overlay.rst` WRT nodejs -> iojs change, in order to have consistent development environments (with and without Docker).

File: Dockerfile [hunk 6]
```diff
@@ -22,6 +22,15 @@
...
+
+# add invenio user
+RUN useradd --home-dir /home/invenio --create-home --shell /bin/bash --uid 1000 invenio
+
+# set work dir
+WORKDIR /code
+
 # nodejs repo
 # detects distribution and adds the right repo
 # See: https://github.com/joyent/node/wiki/Installing-Node.js-via-package-manager#debian-and-ubuntu-based-linux-distributions
```
Comment (tiborsimko) [inline Dockerfile]: The `nodejs` related comment should also be amended after selecting `iojs`.

File: setup.py [hunk 7]
```diff
@@ -119,6 +119,7 @@ def run(self):
     "invenio-documents>=0.1.0.post2",
     "invenio-upgrader>=0.1.0",
     "invenio-unapi>=0.1.0",
+    "invenio-jsonschemas>=0.0.0",
```
Comment (jirikuncar) [inline setup.py]: it's not required dependency

**Comment by tiborsimko** [comment]
Superseded by #3495.

**Pull request closed without merging**